### ISSUE TITLE
Problems & Addition for Readme "How to install/use"

### ISSUE THREAD

**Issue by lukycody** [Naylepsh/guild-wars-2-auto-daily-rewards#5]
Hey there. So I tested quite some stuff and want to give you feedback and problems I encoutnered. 

First off I tried to run this and it turned out for me, that I first needed to install some packages in my vscode:
`py -m pip install pyautogui`
`py -m pip install pyyaml`

May be worth it to point out in the readme description.
Bonus questions just out of curiosity to better understand code:
Why do you need this poetry thing in the first place (packet manager? what does it offer here?) and why do you need to start the main.py script via poetry aswell? (`poetry run python main.py`)

Problem with the script:
Besides that the script is not working for me as it should. The problem is that first the login credentials don't get entered correctly, because the focus/position of the pyautogui is not right. Then the focus swaps to another window (which e.g. is the vscode window) and tries to enter the credentials there, which messes up the whole thing being not in sync.

1. When the launcher starts with previously ticked boxes to save credentials, the focus is on the password field so it seems to mess up the login() function and its credential input.
--> fixed that by deleting previous credentials and unticking the boxes

2. However, the script also tabs out to another window and tries to insert credentials there, not in the launcher fields.
Not sure why its tabbing out at all though, as that behaviour doesnt seem to be in the login() function. Also I'm not sure, why you use double_click() before clearing the input. Maybe this klicking at current position creates the tabbing behaviour (or are you supposed to not even touch the mouse after running the script?)
But I also think that you have to select the correct fields for email and password first (pyautogui.moveTo(x,y)) before trying to clear_input().

Edit: it clearly moves the mouse to a spot where theres no launcher/field to click. My launcher always starts right in the middle of the screen, but the cursor moves to the upper left direction and clicks there. So theres something to fix about the positioning of either the cursor coordinates or the launcher (probably should manually set both positions in the code to sync the actions).

Edit2: it just came to my mind: are the coordinates maybe set for FHD 1920x1080 resolution? I'm using 2560x1440 atm, maybe the reason. If thats the reason you could maybe add a request for a user input of the used resolution at programm start and set the coordinates accordingly to that.
--> that seems to "kinda work" with FHD. Sometimes the mouse position gets on the input field, sometimes it still randomly "tabs out" somewhere else.

Edit3: Another bug in the code;
This doesn't work with mails that have a "+" in them. It gets replaced by "*" which ofc then causes the login to fail.
Idk why it does so, as it should take the original text from the .yaml and just pass it 1:1 as a string not replacing anything.
That actually breaks the whole process if youre using such an email address.

**Comment by Naylepsh** [comment]
About the installation -- there's already a section "How to run" in the README which mentions `poetry install`, which, well, installs the required packages.

About the "Why do you need this poetry thing" -- it's a package manager. IMO it's a better way of handling dependencies than the questionable standard of `venv / pyenv`, `pip` & `requirements.txt` trio.

About the "why do you need to start the main.py script via poetry" -- so that the script has access to the dependencies. Similarly, if you were not to use poetry, but the usual `pip` and `requirements.txt`, before running the script you'd have execute `source ./venv/bin/activate`. Or you could skip this step and install packages globally instead, but that's... not exactly maintainable.

About the "are the coordinates maybe set for FHD 1920x1080 resolution?" -- yes. The README mentions that currently only FHD resolution in supported. Because my screen doesn't support higher resolutions, I'm not able to test other coordinates. However, if you were to input your own and later send a MR request -- that would be appreciated.
And to the related "you could maybe add a request for a user input of the used resolution at programm start and set the coordinates accordingly to that" -- that's a good suggestion, but I need the coordinates for other resolutions first lol.

I'll address the other questions / suggestions / issues later.

**Comment by lukycody** [comment]
Thanks for the quick reply!
Sorry, totally missed that assumption of the resolution! If I find a way to figure out the coordinates of the fields/buttons on screen I‘ll test it on my resolutions up to WQHD! (Can’t test 4K though) and let you know! :)

About the Readme part: I saw that and followed the advice. ‘poetry install‘ didn‘t make it for me though. Executing the script gave the error of missing packages. Somehow, I had to manually install the pyautogui and pyyaml  dependencies via pip, after that it worked.

Only barrier for me now is the replacement of + vs * in the credential text. 
As a workaround it should be possible to build in a custom string slice and replacement step before actually pasting it into the field, but that part with credential reading from the yaml file already was a bit mystically to me me - and I suppose for you too („#noqa“).

**Comment by Naylepsh** [comment]
"(never did that, how did you grab them manually?" -- I just make a screenshot, then open it in paint. Then just hover a cursor over the parts I'm interested in. It displays the coordinates in the bottom left corner.

"Executing the script gave the error of missing packages" -- How have you installed poetry? Through pip or through official installer? The pip way is known to cause some issues sometimes.

"Only barrier for further testing now is the replacement of + vs * in the credential text." -- well, this is weird. I just tried on my end and using "+" in the yaml's email field worked fine lol. Both the extraction from file and passing it to input passed without any issues. Is this happening with any other characters on your end? Or just the "+" -> "*"?

And about the "#noqa" -- that doesn't mean I haven't done QA on my end, it's just a tag for the linting / formatting tools to ignore this line. Otherwise they're gonna complain that the line is too long. But I cannot make it shorter, because lambdas cannot be longer than 1 line.

**Comment by Naylepsh** [comment]
A thought about this "+" -> "*" issue -- can you try to print out your credential's email field in the `login function` and check whether that matches what you have inputted? I wonder whether the extracting data from YAML works fine, but it's `pyautogui.write` that causes the issue.

**Comment by lukycody** [comment]
Ah thats a neat little trick with paint!
Here are the positions/coodinates of the launcher fields (x,y) for WQHD resolution (2560 x 1440):
- email: 760, 730
- pw: 760, 810
- login: 800, 900
- ready/play: 1500, 1000
- first character: tbd

(which field/button is the double_klick here after credential input targeted at?)
`if credentials:
        enter_email()
        enter_password()

    pyautogui.moveTo(x=480, y=720)
    double_click()`


Yep, I did the official install method for windows from their website, just as you stated in the readme.

About the `"+" -> "*"` issue:
It prints the correct data inside `login()`. So there seems to be something wrong with pyautogui then. Any idea how to fix this?
And yes, it's only this one character that gets replaced. So only .+@ used for special characters

#noqa ... I just googled it... quality assurance. 
Oh wow, I thought #noqa would mean "no questions asked" because you also used it first with the following comment in the code: "# For whatever reason double clicking through pyautogui is not enough, but more clicks somehow works # noqa"
It's like "huh whatever, if it, works it works. no questions asked"😂

**Comment by Naylepsh** [comment]
"#noqa" -- Ha, I was looking on a different line when I was thinking about it (get_credentials and not the enforce_click). But the idea's the same -- line is too long.

About the coordinates -- nice, thank you. I'll look into adding those.

And finally about `"+" -> "*"` -- It might be that you're using some non standard keyboard layout. For example, when pressing alt+l normally nothing happens, but when I change the layout by pressing ctrl+shift, then the same alt+l operation now yields `ł` character. Similar thing you might had seen in laptops is than after pressing some functional key, pressing some arbitrary element on keyboard decreases / increases volume or brightness, etc. 
From stack overflow https://stackoverflow.com/questions/73193008/why-is-pyautogui-replacing-with : "This usually happens due to the use of non-qwerty keyboards. When you use write (or press), pyautogui doesn't actually write that character, instead it looks for the associated key on your keyboard and presses it"

**Comment by lukycody** [comment]
Well I‘m using an ISO DE standard keyboard.
Hm, would it suffice to force swap to the EN layout in windows before running the script (or even doing that programmatically) to solve this issue. Or do you actually need the physical periphery for that? That’s weird honestly. It‘d mean, that people with another keyboard other than ansi EN layout can’t reliably use pyautogui 😂

**Comment by Naylepsh** [comment]
Hey, so the `fix-special-char-input` branch https://github.com/Naylepsh/guild-wars-2-auto-daily-rewards/tree/fix-special-char-input should fix the input issues.

**Comment by lukycody** [comment]
Heya, got back to testing.
So it actually works if you swap the language to Englisch/US-Keyboard (bottom right if you have another language set, shortcut: WIN+Space). I added the position for the character selection 1st slot. I couldn't really get you the wiggly chest location, as that's completely reliant on the general UI scale and map size.

I will try your solution aswell, but just fyi that it works that way. How do I get the code's version with that branches addition? (sorry, im not really experienced with those git things yet, have tested with a local copy yet)
What is it actually doing with pyperclip? Doing the workaround with the clipboard?

However I ran into another problem:
- The function `check_for_remind_me_later()` does not work for me. I also tried to print stuff out, but it seems the `pyautogui.locateOnScreen('RemindMeLater_wqhd.png')` actually doesn't find anything on screen (and yes I did make sure to have the launcher actually show it after login in, but even with the .png open on screen it should detect it, which it doesnt).
So for now it works for the case that theres no "remind me" step, otherwise it'll break/desynch at that point.
I even tried to create a new screenshot of the button, just in case its got something to do with the resolution again. But same thing. Prints out `None` aswell.
- (Sometimes the input of `open_gw2_launcher()` is too fast and will result in edge browser search popping up or just nothing happens. Also sometimes the game takes extra long to load, so it'll try to select character while the isnt up yet and breaks the synch again. Just a sleeptimer thing though, or my pc being weird).

Does the picture detection actually work for you?
(Also, why are you not just using `pyautogui.click('RemindMeLater.png')`

**Comment by Naylepsh** [comment]
Assuming you're using git through command line, to test the other branch you simple:
```
- git pull
- git checkout feat-resolutions
```

About pyperclip -- correct, it's just for clipboard handling to enable that previously mentioned workaround.

As for the issues:
1. I have not personally tested `check_for_remind_me_later` functionality -- it's something another dev brought. Because the "Remind me later" doesn't appear in my launcher, I didn't even notice when / whether it stopped working. If you can tell me how to have this "Remind me later" show in the launcher, I can take a look into it. (Assuming that it's a priority of some sort, cause I'm guessing you didn't have this reminder popup before?)
2. Yeah... it's been optimized for my PC, hence the sleep times may require some custom tweaking around on other environments. If you have some specific sleep times that you'd like to use, you can send me the list and I'll create a preset of that.

**Comment by lukycody** [comment]
Thanks, will try it with git.

`remind me later` -- it appears at every first attempt of the day per account and then roughly about every 2nd time, when you login to an account, that doesn't have the two-factor-authentication active. If I click it away once, it won't appear atleast on the next attempt but instantly let you launch. It can pop up again on other attempts after though. 

So yeah, it constantly comes up so its quite important/of priority to get rid off of it, if you wanna automate the logins, as most alt accounts will have that reminder step blocking the launch button.

Would be nice if you could take a look into it. As I said I tested quite a bit but couldn't find the issue.

**Comment by Naylepsh** [comment]
Okay, so I tried it on my end and it looks like it doesn't work for me either. But when I've updated the screenshot it started working, so I'll be committing the new screenshot. However, I'm concerned whether relying on screenshot will work the same in the higher resolutions... 

Fyi: I've included the newest image in the most recent version of https://github.com/Naylepsh/guild-wars-2-auto-daily-rewards/tree/feat-resolutions branch

**Comment by lukycody** [comment]
Ok, will try it with the new screenshot.
However that’s weird, because I tried with a new screenshot on my resolution aswell and it didn‘t work. How did you take the sc, via Snipping Tool or windows print function? I wonder if it’s actually this ‘pixel-perfect‘ requirement.

**Comment by Naylepsh** [comment]
I've used snipping tool and then save through paint.

**Comment by lukycody** [comment]
I tried it with your current branch feat-resolutions (took me a bit to make it running on another computer with FHD, just btw) 
-- it doesn't write the "@" of the email address, always resulting in failed logins. Thus I couldn't test further.

**Comment by Naylepsh** [comment]
Oof. I should had figured that more characters like that may not be inputted properly on a different keyboard layout...

I've updated the branch -- now the entire input for credentials is handled with the "hack" to hopefully make it layout agnostic. The issue of entering the program's name has also been addressed -- a small delay has been added. 

Please let me know whether this update has resolved the aforementioned issues on your end.

**Comment by lukycody** [comment]
Yep, it works now. Thanks! Thats for FHD though, will test with WQHD aswell.
I had a look at the code again and it seems you changed quite a lot about the code structure with the pyperclip/resolution changes (outsourced packages/modules).
It was a bit harder to understand for me with those additional modules and restructured code, but I guess its also a bit cleaner. Do you plan to put this into the main branch?

**Comment by lukycody** [comment]
Alright. Yeah sure, I'll do that. I'm just a bit short on time atm and just on my laptop.
As I said before, those coordinates are not really reliable, as they heavily depend on your ingame setup (e.g. map position up/down and its size). But I'll try to get the coords for the "standard out of the box" setup I guess. Maybe it's also possible to get a average value top- and bottom-y, thats valid for mapsize(min,max), will check.

Also, to learn, I'd come back to you for some code-understanding questions for this project if that's fine :)

**Comment by Naylepsh** [comment]
About the "they heavily depend on your ingame setup" -- yeah, these were initially hardcoded to suit my set-up. The whole ` generate_locations` function is supposed to accommodate for some variance in map size, but it's definitely far from perfect. Would greatly appreciate additional coords from your perspective.


And about the "Also, to learn, I'd come back to you for some code-understanding questions for this project if that's fine" -- no problem at all. By all means ask.

**Comment by lukycody** [comment]
Somehow my launcher seems to have moved a few pixels down and right "overnight", so the coordinates of the email/pw field aren't correct anymore. 
Sure I could manually change them to make it fit. But I'd rather like to understand why it did this and prevent more launcher wiggles. It seems to always start at the same fixed position (if you drag somewhere else the position isn't saved by the launcher/os). And I didn't change anything from yesterday to today, so I wonder if its position somehow got changed the minor game updates. Do you have the same issue today on FHD?

**Comment by Naylepsh** [comment]
Huh, it seems to moved on my end as well... guess that's really due to the update.

**Comment by lukycody** [comment]
Quick Git question :)
So I created my own "personal branch" (named "custom"), just as you've suggested above.
Then I've made local changes to the code there (just added and changed minor stuff like a randomizer for the sleeptimers, which I do not want to merge in main in order push it to the repository). 

Now I want to "update" my personal branch with your new changes, so that I have the latest version (changes coords/.png), BUT also keep my personal changes/additions. What is the right way to do it? (internet throws around different methods like fetch and merge, but I don't want to accidently discard my changes by overwriting)

**Comment by Naylepsh** [comment]
I don't think you can avoid manually resolving conflicts in this scenario. I'd make a backup branch with your local changes just to be on the safe side, so something like:
```
git checkout feat-resolution
git pull
git checkout custom
git checkout -b custom-backup
git checkout custom
git merge feat-resolution
... <- the part where you manually resolve conflicts
git add ...
git commit -m ...
```
This won't magically discard your changes. Worst case scenario you've made a mistake while manually resolving conflicts, but you still have the backup branch to retry.

**Comment by lukycody** [comment]
Heya @Naylepsh.
I know it's been some time and I wanted to get back to you, Sorry for that. Been busy but I didn't forget.

So first off I just wanted to let you know that I've been using the tool for quite some time now and - once correctly set up with the right pause timers for your machine and besides _some shenanigans_ - it worked great so far to login daily with several accounts. 
- For ease of use/startup I created an executable file with the target: `"C:\Windows\System32\cmd.exe /k cd C:\Users\...\guild-wars-2-auto-daily-rewards && echo [Starting Programm...] && poetry run python -m main -r FHD"`
- Also I customized the code a bit besides the pause timers and deactivated the function for clicking the daily boxes -- as thats not really needed tbh (they stack up in overflow and have to be claimed alltogether at some point anyway).
- I noticed that the login takes longer incrementally each attempt and ramps up to make login attempts fail. So I added a ~5min pause after 20 iterations/accounts, as that seems to be the sweetspot, at that a pause is required because the login duration becomes too long to finish and instead fails (which then breaks the script). 
- As a sidenote, iirc the maximum amount of different account logins per IP seems to be 40 (taken from YT MightyTeapots video about that). VPNs should be avoided, as that could lead to false positives for malicious attempt detection in the backend.

About the shenanigans -- I sometimes encounter the problem, that the script will stop (or rather desync for whatever reason) after the 5min pause window. So it indeed runs the script to the end but the actually correct execution of logging in stopped after the 5min pause (I could verify that by starting GW2 normally and seeing the last saved account (number 21) credentials in the launcher)
A thing I noticed is, that in the CLI (where I print out "Account [i] started/finished" for each iteration) at the end of the script (i print out "program finished" at the very end) that a part of the printed out text in the middle of the cmd window is highlighted/marked, so at some point the clicking was desynced and happened with focus on something else than the gw2 launcher window. Couldn't reproduce/see it in real time when it happens yet.
It doesn't happen always, but apparently randomly. One reason I could think of is that maybe during the pause windows gives other programs in the background ressource priorities and it somehow breaks the focus on the program. 

Okay but that aside theres currently a problem with the copy&paste of account credentials that stops the script from working -- atleast for me, can you confirm it on your end?
So when it copies the account credentials into the email and password fields, instead of
- clicking into the field, then selecting all (Strg+A), then copying the whole email/password, 
... instead it:
- clicks into the field
- copies in just one symbol: » 
([U+00BB](https://util.unicode.org/UnicodeJsps/character.jsp?a=00BB), [&#187;](https://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references)
and tries to move on from that.

Happens for me after the update on Tuesday. However, I'm not entirely certain, if thats necessarily connected to the game update or rather something else like the pyperclip dependency/plugin

**Comment by lukycody** [comment]
The coordinates of the daily boxes positions in WQHD resolution are as follows (x,y)
I assumed a larger UI size with WQHD, because else the UI gets really small, so its unlikely people with use that with this resolution.

with map UI on top:
- UI size large: 2500, 1390
- UI size larger: 2490, 1380
with map ui on bottom (assuming max. sized map)
- UI size large: 2500, 970
- UI size larger 2490, 920

**Comment by lukycody** [comment]
Hey, sorry to disturb, but do you have any idea how to fix this problem? I had no luck fixing it yet.

**Comment by Naylepsh** [comment]
It looks like an issue with the launcher. When I tried running the script to output the data to notepad instead of the launcher (you can do the same by replacing "Guild Wars 2" with "notepad":
```python
def _open_gw2_launcher() -> None:
    _open_program("notepad")
```
and it pasted the credentials just fine

**Comment by Naylepsh** [comment]
The dumb solution would be to go back to the naive way of typing the credentials through pressing the keys:
```python
def write_input(xs: str) -> None:
  for x in xs:
     pyautogui.press(x)
```
but that won't quite work with the non-standard symbols (as we had this issue before, hadn't we)

**Comment by lukycody** [comment]
Well that `pyautogui.press("backspace")` didn't even work as it put that arrow symbol... so a thousand times wont work aswell 😀
So it seems we don't have a working solution/workaround at the moment huh 🤔

By migrating you mean writing the code anew in Java?

**Comment by Naylepsh** [comment]
I tried a bunch yesterday but they didn't work at all. I found https://pypi.org/project/pynput/ which seems to do the job (at least when it comes to backspace), so maybe there's hope

**Comment by Naylepsh** [comment]
Alright, so the [feat-resolutions branch](https://github.com/Naylepsh/guild-wars-2-auto-daily-rewards/tree/feat-resolutions) should cover the fixes.

**Comment by lukycody** [comment]
Nice, thank you! Seems to work for me after adding my custom changes (timers) again. Before I pulled I stashed them away and applied them after.

Some more on it though:
* I had to renew the "RemindMeLater" picture again on FHD to let it be recognized again.
* I noticed you still had the most of the flow.py/run function still commented out.
* Used poetry install again to get the pynput dependency. Getting some message on the imports though: 
![image](https://github.com/Naylepsh/guild-wars-2-auto-daily-rewards/assets/63720164/86ef297f-5c08-4bb2-b754-154b477874bc)

Another thing (that already happened before) is that the script stops working after the 5 minute break after 20 accounts, that I built in to prevent the login attempt timeout.
Sometimes it runs through normally and continues with #21, sometimes it seems to break focus and end up with the cursor somewhere in the cmd window instead of the game launcher.

**Comment by Naylepsh** [comment]
> I had to renew the "RemindMeLater" picture again on FHD to let it be recognized again.

Yeah, I've noticed that it tends to break on updates, I don't think there's much I can do about that.


> I  noticed you still had the most of the flow.py/run function still commented out.

Nice catch, completely forgot about it. 😆 


> Used poetry install again to get the pynput dependency. Getting some message on the imports though

It's just your IDE not using the correct environment.


> Sometimes it runs through normally and continues with #21, sometimes it seems to break focus and end up with the cursor somewhere in the cmd window instead of the game launcher


Unfortunately I have no definite clue what's causing it. It could be that every now and then the launching gw2 launcher takes longer than the expected delay and so the login part starts executing on the underlying cmd window instead of the gw2 launcher. Or it could be that the launcher shows in the background instead of foreground (had such a thing happen I think once when I was testing the recent fix) 🤷 
You might want to try out extending the sleep time between `_open_gw2_launcher` and `_login` and see what happens
